### PR TITLE
[JS] Topic: Async and Await

### LINKED ISSUE #721: JavaScript: What is Async and Await?
Description: I Will be explaining what is Async and Await in JavaScript (ES6).

The async keyword turns a method into an async method, which allows you to use the await keyword in its body. When the await keyword is applied, it suspends the calling method and yields control back to its caller until the awaited task is complete. await can only be used inside an async method.
Changes will be inside winter-of-contributing/Web_Development/Javascript/Topics/ branch.

I am a GWOC'21 contributor from Batch 1 of JavaScript. Please assign this issue to me. I assure you that this issue's title will be the same as my PR's title as well as I will be Contributing under the Guidelines and Code of Conduct.

### PR DESCRIPTION
**Pull request by Diksha-11**
<hr>

## Added Async and Await documentation.
Added Async and Await documentation in the Javascript branch, fixes the Async and Await issue 

The async keyword turns a method into an async method, which allows you to use the await keyword in its body. When the await keyword is applied, it suspends the calling method and yields control back to its caller until the awaited task is complete. await can only be used inside an async method.

Fixes #721 

<hr>

## Type of change 📝

- [ ] Au#dio (Should be in mp3 format Includes speech clarity, Concise,Low distortion)
- [ ] Vi#deo (Animations, screen-recordings, presentations, and regular explanatory films are all possibilities, etc)
- [X] Doc#umentation (Content Creation in the form of codes or tutorials)
- [ ] Other (If you choose other, Please mention changes below) 

<hr>

## Domain of Contribution 📊

- [ ] Android Dev #(Flutter)
- [ ] Android Dev #(Java)
- [ ] Android Dev #(Kotlin)
- [ ] Backend Dev #(Java)
- [ ] Backend Dev #(.NET)
- [ ] Backend Dev #(PHP)
- [ ] Backend Dev #(Python)
- [ ] C/#CPP
- [ ] Competitive #Programming
- [ ] Cyber #Security
- [ ] DS#A
- [ ] Data#base
- [ ] Datascience with #Python
- [ ] Datascience with #R
- [ ] Frontend Dev #HTML/CSS/JS
- [ ] Frontend Dev #React/Angular/Vue
- [ ] Go#lang
- [ ] Interview #Prep
- [ ] Java_#Domain
- [ X ] Java#Script
- [ ] ME#RN
- [ ] Machine #Learning
- [ ] Open #Source
- [ ] Python_#Domain
- [ ] Ru#st
- [ ] Statis#tics
- [ ] UI/#UX

<hr>

## Checklist ✅

- [ X ] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [ X ] I have performed a self-review of my own code or work.
- [ X ] I have commented my code, particularly in hard-to-understand areas.
- [ X ] My changes generates no new warnings.
- [ X ] I'm GWOC'21 contributor

<hr>

## Screenshots / Gif (Optional) 📸

<hr>

### CODE REVIEW

**Comment by Diksha-11** [comment]
Please review my PR @paramsiddharth

**Comment by paramsiddharth** [comment]
@Creatoon @dakshsethi @s4-2 Please review.

File: JavaScript/Topics/Async_Await/README.md [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    }, randomNumberOfSeconds * 1000);
+  });
+});
+ 
+ 
+ HAPPY LEARNINGS!
+ 
+ #### Contributed by 
+ [Diksha Sharma]
+ ( https://github.com/Diksha-11 )
```
Comment (paramsiddharth) [inline JavaScript/Topics/Async_Await/README.md:353]:
```suggestion
 [Diksha Sharma](https://github.com/Diksha-11)
```

**Review by paramsiddharth — CHANGES_REQUESTED** [review]
Please enclose all code-blocks like the following:
````
```js
async function doSomeThingHere() {
    return await iDidSomeThingHere();
}
```
````

This results in:
```js
async function doSomeThingHere() {
    return await iDidSomeThingHere();
}
```

**Comment by paramsiddharth** [comment]
@Diksha-11 Make sure you attend our acknowledgement sessions on Saturdays! :) Keep in touch with the team on the general channel.

**Comment by Diksha-11** [comment]
> @Diksha-11 Make sure you attend our acknowledgment sessions on Saturdays! :) Keep in touch with the team on the general channel.

Sure sir!

**Comment by s4-2** [comment]
@Diksha-11 Please keep only one Fixes #721  in the Description section.

**Comment by Diksha-11** [comment]
> @Diksha-11 Please keep only one Fixes #721 in the Description section.

Done mam. I've edited the Description section.

**Comment by Diksha-11** [comment]
@s4-2 @paramsiddharth what changes are requested??


> Please enclose all code-blocks like the following:
> 
> ```
> ```js
> async function doSomeThingHere() {
>     return await iDidSomeThingHere();
> }
> ```
> ```
> 
> This results in:
> 
> ```js
> async function doSomeThingHere() {
>     return await iDidSomeThingHere();
> }
> ```

File: JavaScript/Topics/Async_Await/README.md [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+4. Better Error Handling.
+
+A Promise has four states:-
+ 
+1. fulfilled: Action related to the promise succeeded
+2. rejected: Action related to the promise failed
+3. pending: Promise is still pending i.e not fulfilled or rejected yet
+4. settled: Promise has fulfilled or rejected
+
+ ~~EXAMPLE 1:
```
Comment (Diksha-11) [inline JavaScript/Topics/Async_Await/README.md:36]:
```js

File: JavaScript/Topics/Async_Await/README.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
+# Async and Await
+
+JavaScript Introduces keywords “async” and “await” as an addition to EcmaScript 2015.We are going to see about the 
+following concepts.
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:4]: why this is next line

File: JavaScript/Topics/Async_Await/README.md [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+1. Working With Async Keyword
+2. Impact of Await Keyword
+
+## What is Async and Await?
+
+Before getting in details of Async and Await, We should have a good understanding of Promises in JavaScript. 
+
+Async/Await is the extension of promises which we get as a support in the language. So before going to understand what
+is Async/Await, We should, first understand Promises in Javascript.
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:14]: again in next line, aren't they should be in same line

File: JavaScript/Topics/Async_Await/README.md [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+
+### What is Promises in JavaScript?
+
+Promises are used to handle asynchronous operations in JavaScript.
+Promises are the ideal choice for handling asynchronous operations in the simplest manner. They can handle multiple 
+asynchronous operations easily and provide better error handling than callbacks and events.
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:20]: next line issue,

Please remove these things from the whole doc if it contains, i am not marking this further

File: JavaScript/Topics/Async_Await/README.md [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+ ~~EXAMPLE 1:
+let myPromise = new Promise(function(myResolve, myReject) 
+{
+  myResolve(); // when successful
+  myReject();  // when error
+});
+
+myPromise.then(
+  function(value) { /* code if successful */ },
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:44]:
```suggestion
  function(result) { /* code if successful */ },
```

result is the correct name for the fullfilled state arg

File: JavaScript/Topics/Async_Await/README.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+  function(error) { /* code if some error */ }
+);
+
+Success:	myResolve(result value)
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:48]:
```suggestion
Success: myResolve(result value)
```

File: JavaScript/Topics/Async_Await/README.md [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+is resolved with its value.
+
+So, Returning Promises with Async Keyword: 
+When we append the keyword “async” to the function, this function returns the Promise by default on execution. Async 
+keyword provides extra information to the user of the function:
+1. The function contains some Asynchronous Execution
+2. The returned value will be the Resolved Value for the Promise.
+
+~~EXAMPLE 1: 
+async function simpleAsync() {
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:66]: put the code inside js editor

File: JavaScript/Topics/Async_Await/README.md [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+  return "1"
+}
+
+simpleAsync().then((data) => {
+    console.log(data)
+})
+
+OUTPUT: 1
+
+To in order to summarize the impact of the “Async” keyword, we can say that:
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:89]:
```suggestion
In order to summarize the impact of the “async” keyword, we can say that:
```

File: JavaScript/Topics/Async_Await/README.md [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+1. Appending “async” will return a promise from a function
+2. In case the function is returning some value, the value will be available as the resolved value from the promise.
+3. If no value is returned, the resolved value will be “undefined”
+
+To see the whole Asynchronous behaviour:
+~~EXAMPLE 3:-
+
+function returnPromises() {
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:98]: put all functions in the javascript editor, use proper markdown techniques

File: JavaScript/Topics/Async_Await/README.md [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+}
+
+ExecuteFunction()
+
+OUTPUT:  Mayank
+         Undefined.
+		 Promise Executed...
+
+
+## Await:
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:120]:
```suggestion
## await:
```

write all the keywords in correct format, same for Async -> async

File: JavaScript/Topics/Async_Await/README.md [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+}
+  
+console.log(1);
+getData();
+console.log(2);
+
+
+OUTPUT: 1
+        2
+Hello World
```
Comment (Creatoon) [inline JavaScript/Topics/Async_Await/README.md:138]: indent these properly

**Review by Creatoon — CHANGES_REQUESTED** [review]

**Comment by Creatoon** [comment]
@Diksha-11 please make changes to the whole doc, not only to the specified places i made, there are lots of changes to be done before getting merged

File: JavaScript/Topics/Async_Await/README.md (hunk 5 above)
Comment (Diksha-11) [reply JavaScript/Topics/Async_Await/README.md:20]: @Creatoon sir,
I prepared this documentation in notepad++. while reading the docs, there was a scroll at the bottom that's why I broke the line into several for convenience but forgot to remove those extra lines while making PR. My Fault. 
Thank you for the review. I will correct it.

**Review by dakshsethi — APPROVED** [review]
Looks good to me!! 🔥🔥🔥

**Comment by Diksha-11** [comment]
@dakshsethi Thanks, Sir.

File: JavaScript/Topics/Async_Await/README.md [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+      }
+    }, randomNumberOfSeconds * 1000);
+  });
+});
+ ```
+ 
+ HAPPY LEARNINGS!
+ 
+ ##Contributed by, 
+ [Diksha Sharma] (https://github.com/Diksha-11)
```
Comment (paramsiddharth) [inline JavaScript/Topics/Async_Await/README.md:324]:
```suggestion
 ## Contributed by, 
 [Diksha Sharma] (https://github.com/Diksha-11)
```
This way, it renders correctly.

**Review by paramsiddharth — CHANGES_REQUESTED** [review]

**Comment by paramsiddharth** [comment]
@Creatoon @dakshsethi @s4-2 Please review.

**Comment by paramsiddharth** [comment]
@Creatoon @s4-2 Please review.

File: JavaScript/Topics/Async_Await/README.md [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+      }
+    }, randomNumberOfSeconds * 1000);
+  });
+});
+ ```
+ 
+ HAPPY LEARNINGS!
+ 
+ ## Contributed by, 
+ [Diksha Sharma] (https://github.com/Diksha-11)
```
Comment (paramsiddharth) [inline JavaScript/Topics/Async_Await/README.md:324]:
```suggestion
 [Diksha Sharma](https://github.com/Diksha-11)
```
Else it won't render. Please fix! Remove the space between `]` and `(`.

**Review by paramsiddharth — CHANGES_REQUESTED** [review]

**Comment by paramsiddharth** [comment]
Thank you so much for the amazing PR, Diksha! Sadly another PR for your issue (which was created before yours) under documentation has been merged. If you wish to contribute further to the same issue, you can create a PR improving the merged document or add your contribution in another format. Additionally, you can create multiple PRs for different issues, or create your own based on the roadmap here: https://github.com/paramsiddharth/gwoc-js-roadmap/releases/latest/download/JavaScript.Roadmap.for.GWoC.pdf

Good luck! :) Do join our acknowledgement session on Saturdays, 7 PM! We'd love to see you there.

**Pull request closed without merging**